hazelcast/cp.py

```diff
@@ -152,3 +200,203 @@ def _get_group_id(self, proxy_name):
         invocation_service = self._context.invocation_service
         invocation_service.invoke(invocation)
         return invocation.future.result()
+
+
+class _SessionState(object):
+    __slots__ = ("id", "group_id", "ttl", "creation_time", "acquire_count")
+
+    def __init__(self, state_id, group_id, ttl):
+        self.id = state_id
+        self.ttl = ttl
+        self.group_id = group_id
+        self.creation_time = time.time()
+        self.acquire_count = AtomicInteger()
+
+    def acquire(self, count):
+        self.acquire_count.add(count)
+        return self.id
+
+    def release(self, count):
+        self.acquire_count.add(-count)
+
+    def is_valid(self):
+        return self.is_in_use() or not self._is_expired(time.time())
+
+    def is_in_use(self):
+        return self.acquire_count.get() > 0
+
+    def _is_expired(self, timestamp):
+        expiration_time = self.creation_time + self.ttl
+        if expiration_time < 0:
+            expiration_time = six.MAXSIZE
+        return timestamp > expiration_time
+
+    def __eq__(self, other):
+        return isinstance(other, _SessionState) and self.id == other.id
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+_NO_SESSION_ID = -1
+
+
+class ProxySessionManager(object):
+    def __init__(self, context):
+        self._context = context
+        self._mutexes = dict()  # RaftGroupId to RLock
+        self._sessions = dict()  # RaftGroupId to SessionState
+        self._thread_ids = dict()  # (RaftGroupId, thread_id) to global thread id
+        self._heartbeat_timer = None
+        self._shutdown = False
+        self._lock = RLock()
+
+    def get_session_id(self, group_id):
+        session = self._sessions.get(group_id, None)
+        if session:
+            return session.id
+        return _NO_SESSION_ID
+
+    def acquire_session(self, group_id, count):
+        return self._get_or_crate_session(group_id).continue_with(lambda state: state.result().acquire(count))
+
+    def release_session(self, group_id, session_id, count):
+        session = self._sessions.get(group_id, None)
+        if session and session.id == session_id:
+            session.release(count)
+
+    def invalidate_session(self, group_id, session_id):
+        session = self._sessions.get(group_id, None)
+        if session and session.id == session_id:
+            self._sessions.pop(group_id, None)
```

Comment:
Good catch. However, after looking for the usages of this function, I see that it is only called from the reactor thread. Since the one thread that modifies the `self._sessions` is the reactor, this shouldn't cause any problems

Reply:
Good to know. Thanks.